### PR TITLE
Added BalanceComputation reports

### LINKED ISSUE #105: BalanceComputation reports
### Describe the current behavior

The `BalancedComputation.run` method doesn't have a reporter parameter.

### Describe the expected behavior

Add a reporter parameter, also produce balance computation reports.

### Describe the motivation

Auditability of balance computation results

### Extra Information

n/a

### PR DESCRIPTION
**Pull request by georgebudau**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

yes, resolves #105 

**Does this PR introduce a breaking change or deprecate an API?**
No

### CODE REVIEW

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java [hunk 1]
```diff
@@ -61,6 +62,14 @@ public BalanceComputationImpl(List<BalanceComputationArea> areas, ComputationMan
...
+    }
+
+    /**
+     * Run balances adjustment computation in several iterations
+     */
+    @Override
+    public CompletableFuture<BalanceComputationResult> run(Network network, String workingStateId, BalanceComputationParameters parameters, Reporter reporter) {
         Objects.requireNonNull(network);
         Objects.requireNonNull(workingStateId);
         Objects.requireNonNull(parameters);
```
Comment (obrix) [inline balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java:75]: you can also do Objects.requireNonNull(reporter)

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java [hunk 2]
```diff
@@ -82,7 +91,7 @@ public CompletableFuture<BalanceComputationResult> run(Network network, String w
             });
 
             // Step 2: compute Load Flow
-            LoadFlowResult loadFlowResult = loadFlowRunner.run(network, workingVariantCopyId, computationManager, parameters.getLoadFlowParameters());
+            LoadFlowResult loadFlowResult = loadFlowRunner.run(network, workingVariantCopyId, computationManager, parameters.getLoadFlowParameters(), reporter);
```
Comment (obrix) [inline balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java:94]: Can you create a subReporter for the load flow (reporter.createSubReporter) ? This way the load flow part will be explictly separated in the report as more generic balance computation reports could be added later. (see how it is done for example here : https://github.com/powsybl/powsybl-open-loadflow/blob/323d052eeab1026f769addfca7a151a49d51855b/src/main/java/com/powsybl/openloadflow/sa/AcSecurityAnalysis.java#L170)

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java (hunk 1 above)
Comment (georgebudau) [reply balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java:75]: done

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java (hunk 2 above)
Comment (georgebudau) [reply balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java]: @obrix I've created a subReporter, could you please check if it's implemented properly?

**Review by obrix — APPROVED** [review]

**Review by obrix — APPROVED** [review]

**Review by jeandemanged — CHANGES_REQUESTED** [review]

File: balances-adjustment/src/test/resources/balancedNetworkReport.txt
Comment (jeandemanged) [inline balances-adjustment/src/test/resources/balancedNetworkReport.txt:1]: as discussed through MP, use subreporters to highlight the processing

I was thinking of something like this

```
 + Balance Computation iteration 0
   + Load Flow
       ....
   + Mismatch
        Mismatch for area XXX: AAAA (target=BBBBB, balance=CCCC)
        .....

+ Balance Computation iteration 1
   + Scaling
       Scaling for area XXX: offset=AAAA , done=BBBBB
       ...
   + Load Flow
       ....
   + Mismatch
        .....
 
+ Balance Computation iteration 2
   + Scaling
       .....
   + Load Flow
       ....
   + Mismatch
        .....
   + Termination
          Areas [A1, A2, ...] are balanced after N iterations

```

**Review by jeandemanged — CHANGES_REQUESTED** [review]

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java [hunk 3]
```diff
@@ -61,9 +63,18 @@ public BalanceComputationImpl(List<BalanceComputationArea> areas, ComputationMan
...
+     * Run balances adjustment computation in several iterations
+     */
+    @Override
+    public CompletableFuture<BalanceComputationResult> run(Network network, String workingStateId, BalanceComputationParameters parameters, Reporter reporter) {
         Objects.requireNonNull(network);
         Objects.requireNonNull(workingStateId);
         Objects.requireNonNull(parameters);
+        Objects.requireNonNull(reporter);
 
         BalanceComputationRunningContext context = new BalanceComputationRunningContext(areas, network, parameters);
```
[the remark this answers is not in the document]
Comment (georgebudau) [reply balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java]: done

File: balances-adjustment/src/test/resources/balancedNetworkReport.txt
Comment (jeandemanged) [reply balances-adjustment/src/test/resources/balancedNetworkReport.txt:1]: that's great, thanks!

**Review by jeandemanged — APPROVED** [review]

File: balances-adjustment/src/test/resources/balancedNetworkReport.txt
Comment (georgebudau) [reply balances-adjustment/src/test/resources/balancedNetworkReport.txt:1]: thank you ❤️

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java [hunk 4]
```diff
@@ -74,27 +85,34 @@ public CompletableFuture<BalanceComputationResult> run(Network network, String w
...
             context.getBalanceOffsets().forEach((area, offset) -> {
                 Scalable scalable = area.getScalable();
                 double done = scalable.scale(network, offset, parameters.getScalingParameters());
+                Reports.reportScaling(scalingReporter, area.getName(), offset, done);
                 LOGGER.info("Iteration={}, Scaling for area {}: offset={}, done={}", context.getIterationNum(), area.getName(), offset, done);
             });
 
             // Step 2: compute Load Flow
-            LoadFlowResult loadFlowResult = loadFlowRunner.run(network, workingVariantCopyId, computationManager, parameters.getLoadFlowParameters());
+            Reporter lfReporter = iterationReporter.createSubReporter("loadflow", "Load Flow");
```
Comment (annetill) [inline balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java:99]: Just for coherence, in OLF we have key "loadFlow", message "Load flow ..."

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java [hunk 5]
```diff
@@ -110,13 +128,17 @@ public CompletableFuture<BalanceComputationResult> run(Network network, String w
             }
         } while (context.getIterationNum() < parameters.getMaxNumberIterations() && result.getStatus() != BalanceComputationResult.Status.SUCCESS);
 
+        Reporter terminationReporter = reporter.createSubReporter("termination", "Termination");
```
Comment (annetill) [inline balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java:131]: Termination?

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/util/Reports.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                .withValue("offset", offset)
+                .withValue("done", done)
+                .withSeverity(TypedValue.INFO_SEVERITY)
+                .build());
+    }
+
+    public static void reportConvergenceError(Reporter reporter) {
+        reporter.report(Report.builder()
+                .withKey("convergenceError")
+                .withDefaultMessage("LoadFlow on network does not converge")
```
Comment (annetill) [inline balances-adjustment/src/main/java/com/powsybl/balances_adjustment/util/Reports.java:42]: "Load flow" for coherence. Note that in general a load flow does not converged on a synchronous component number with open-loadflow. You must be more accurate here.

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java (hunk 5 above)
Comment (jeandemanged) [reply balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java:131]: my bad, `Status` would probably be better

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/util/Reports.java (hunk 6 above)
Comment (jeandemanged) [reply balances-adjustment/src/main/java/com/powsybl/balances_adjustment/util/Reports.java:42]: we should probably put the iterationReporter in the context, and move the loadflow subreport creation inside the `isLoadFlowResultOk` method

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java (hunk 4 above)
Comment (georgebudau) [reply balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java]: Done

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java (hunk 5 above)
Comment (georgebudau) [reply balances-adjustment/src/main/java/com/powsybl/balances_adjustment/balance_computation/BalanceComputationImpl.java]: Done

File: balances-adjustment/src/main/java/com/powsybl/balances_adjustment/util/Reports.java (hunk 6 above)
Comment (georgebudau) [reply balances-adjustment/src/main/java/com/powsybl/balances_adjustment/util/Reports.java]: @annetill @jeandemanged 

I've added the status of the convergence of the load flow, could you please check if it's in the right format?

**Review by jeandemanged — APPROVED** [review]

**Review by obrix — APPROVED** [review]

**Pull request merged**